File: algorithms/cpp/212.单词搜索-ii.cpp
#include <iostream>
#include <set>
#include <string>
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=212 lang=cpp
 *
 * [212] 单词搜索 II
 */

// @lc code=start
class Trie {
 private:
  bool isEnd;
  Trie* next[52];

 public:
  Trie() {
    isEnd = false;
    memset(next, 0, sizeof(next));
  }
  int get_ch(const char& ch) {
    return ch - 'a' > -1 ? ch - 'a' + 26 : ch - 'A';
  }
  void insert(string word) {
    Trie* head = this;
    for (auto ch : word) {
      if (head->next[get_ch(ch)] == nullptr) {
        Trie* node = new Trie();
        head->next[get_ch(ch)] = node;
      }
      head = head->next[get_ch(ch)];
    }
    head->isEnd = true;
  }
  bool search(string word) {
    Trie* head = this;
    for (auto ch : word) {
      if (head->next[get_ch(ch)] == nullptr) return false;
      head = head->next[get_ch(ch)];
    }
    return head->isEnd;
  }
  bool startsWith(string word) {
    Trie* head = this;
    for (auto ch : word) {
      if (head->next[get_ch(ch)] == nullptr) return false;
      head = head->next[get_ch(ch)];
    }
    return true;
  }
};

vector<int> dx = {1, -1, 0, 0};
vector<int> dy = {0, 0, 1, -1};

class Solution {
 private:
  Trie* root;

 public:
  bool in_size(int i, int j, int x, int y) {
    return i < x && i > -1 && j < y && j > -1;
  }
  void dfs(string prefix, vector<vector<char> >& board, set<string>& cur_ans,
           int i, int j, const int row, const int& col) {
    // terminator
    if (!root->startsWith(prefix)) return;
    if (root->search(prefix)) {
      //   cur_ans.push_back(prefix);
      cur_ans.insert(prefix);
    };

    char ch = board[i][j];
    board[i][j] = 0;

    // drill down
    for (int k = 0; k < 4; k++) {
      if (in_size(i + dx[k], j + dy[k], row, col) &&
          board[i + dx[k]][j + dy[k]]) {
        dfs(prefix + board[i + dx[k]][j + dy[k]], board, cur_ans, i + dx[k],
            j + dy[k], row, col);
      }
    }
    // restore state
    board[i][j] = ch;
  }

  vector<string> findWords(vector<vector<char> >& board,
                           vector<string>& words) {
    set<string> cur_set;
    for (auto str : cur_set) {
      cout << str << endl;
    }
    root = new Trie();
    for (auto word : words) {
      root->insert(word);
    }
    int row = board.size(), col = board[0].size();
    for (int i = 0; i < board.size(); i++) {
      for (int j = 0; j < board[i].size(); j++) {
        string prefix = string(1, board[i][j]);
        if (root->startsWith(prefix)) {
          dfs(prefix, board, cur_set, i, j, row, col);
        }
      }
    }
    return vector<string>(cur_set.begin(), cur_set.end());
  }
};
// @lc code=end

File: algorithms/golang/212.单词搜索-ii.go
/*
 * @lc app=leetcode.cn id=212 lang=golang
 *
 * [212] 单词搜索 II
 */

// @lc code=start

type Trie struct {
	children [26]*Trie
	isEnd    bool
}

func Constructor() Trie {
	return Trie{}
}

func (this *Trie) Insert(word string) {
	parent := this
	for _, ch := range word {
		node := parent.children[ch-'a']
		if node == nil {
			node = &Trie{}
			parent.children[int(ch-'a')] = node
		}
		parent = node
	}
	parent.isEnd = true
}

func (this *Trie) Search(word string) bool {
	parent := this
	for _, ch := range word {
		node := parent.children[ch-'a']
		if node == nil {
			return false
		} else {
			parent = node
		}
	}
	return parent.isEnd
}

func (this *Trie) StartsWith(prefix string) bool {
	parent := this
	for _, ch := range prefix {
		node := parent.children[ch-'a']
		if node == nil {
			return false
		} else {
			parent = node
		}
	}
	return true
}

func findWords(board [][]byte, words []string) []string {
	trie := &Trie{}
	for _, word := range words {
		trie.Insert(word)
	}
	retMap := make(map[string]bool)
	row, col := len(board), len(board[0])

	var helper func(i, j int, word string, cur_trie *Trie)
	dx, dy := []int{1, 0, -1, 0}, []int{0, 1, 0, -1}
	helper = func(i, j int, word string, cur_trie *Trie) {
		ch := board[i][j]
		cur_trie = cur_trie.children[ch-'a']
		if cur_trie == nil {
			return
		}
		newWord := word + string(ch)
		if cur_trie.isEnd {
			retMap[newWord] = true
		}
		board[i][j] = '.'
		for k := 0; k < len(dx); k++ {
			x, y := i+dx[k], j+dy[k]
			if x < 0 || x >= row || y < 0 || y >= col || board[x][y] == '.' {
				continue
			}
			helper(x, y, newWord, cur_trie)
		}
		board[i][j] = ch
	}
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			helper(i, j, "", trie)
		}
	}
	ret := make([]string, 0, len(retMap))
	for k := range retMap {
		ret = append(ret, k)
	}
	return ret
}

// @lc code=end
